pkg/safenetlink/netlink_linux.go

```diff
@@ -0,0 +1,395 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package safenetlink
+
+import (
+	"context"
+	"errors"
+	"net"
+
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netlink/nl"
+
+	"github.com/cilium/cilium/pkg/resiliency"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+const (
+	netlinkRetryInterval = 1 * time.Millisecond
+	netlinkRetryMax      = 30
+)
+
+// withRetry runs the netlinkFunc. If netlinkFunc returns netlink.ErrDumpInterrupted, the function is retried up
+// to 20 times. If success or any other error is returned, withRetry returns immediately, propagating the error.
+func withRetry(netlinkFunc func() error) error {
```

Comment:
Should we have metrics or some other observability tooling to introspect netlink commands that are hitting high churn, maybe count when the number of retries is high or something? I'm thinking that if Cilium is actively hitting an issue with one of these commands that causes either undue delay in configuring the kernel / dumping state or such then this would be nice to know as a hint if we experience any issues with some subsystem in a production environment.

Reply:
>  The main question to me would be whether this could cause significant delay in executing some netlink command deep in some callstack, maybe leading to other problems in this adverse case. On that front I raised a question about metrics below.

With the current parameters, the maximum delay is 50 milliseconds. After that, the call unblocks and the wrapper returns an error.


> Should we have metrics or some other observability tooling to introspect netlink commands that are hitting high churn, maybe count when the number of retries is high or something? 

I'll see if that is easily doable. One issue is that the wrappers are also used by cilium-cni (which is actually the main motivation for this PR, since we observed issues with CNI ADD) and in those wrappers the metrics won't work.

Reply:
I attempted to add metrics, but it ends up being kinda ugly and complicated due to 

a) the fact that I want the `safenetlink` to act as a drop-in replacement for the `netlink` calls and thus having to use global variables to store the counters
b) the fact that the existing `netlink` is transitively used by the `metrics` package (via pkg/option) :dizzy_face: Breaking this cycling dependency requires some shuffling of code that I'd like to avoid doing in this PR (that we should backport)